### PR TITLE
Crash fix cudaMallocAsync usage of TF_CUDA_MALLOC_ASYNC_SUPPORTED_PREALLOC.

### PR DESCRIPTION
**Pull request by nouiz**
Some of the refactoring caused crash when using TF_CUDA_MALLOC_ASYNC_SUPPORTED_PREALLOC.
This PR fix them and add a test to be sure it continue to work.

This PR is on top of https://github.com/tensorflow/tensorflow/pull/50961 as otherwise there would be code conflict.

Now, GpuCudaMallocAsyncAllocator doesn't create its own compute stream. But it use one stream set after construction when GpuCudaMallocAsyncAllocator::SetStream is called. So it is now impossible to preallocate all memory during the object construction.
So this PR postpone the memory preallocation to when the stream is available, in SetStream.

This PR also check that the stream isn't set when SetStream is called.

@sanjoy

### CODE REVIEW

**Review by sanjoy — APPROVED** [review]

**Comment by sanjoy** [comment]
Please make the PR description a bit more descriptive (what was the bug?).

CC @hanbinyoon

**Comment by nouiz** [comment]
I updated the description.

File: tensorflow/core/common_runtime/gpu/gpu_cudamallocasync_allocator.cc [hunk 1]
```diff
@@ -374,4 +359,40 @@ bool GpuCudaMallocAsyncAllocator::ClearStats() {
   return true;
 }
 
+void GpuCudaMallocAsyncAllocator::SetStream(void* stream) {
```
Comment (sanjoy) [inline tensorflow/core/common_runtime/gpu/gpu_cudamallocasync_allocator.cc:362]: The function is now misleadingly named -- maybe call it `SetStreamAndPreallocateMemory`?

**Review by sanjoy — CHANGES_REQUESTED** [review]

**Comment by gbaned** [comment]
@nouiz Can you please check @sanjoy's comments and keep us posted ? Thanks!

File: tensorflow/core/common_runtime/gpu/gpu_cudamallocasync_allocator.cc (hunk 1 above)
Comment (nouiz) [reply tensorflow/core/common_runtime/gpu/gpu_cudamallocasync_allocator.cc]: I rebased the PR and changed the function name.
I'm not sure I like it better then the short name.

Note, this PR depends on https://github.com/tensorflow/tensorflow/pull/50961 to be merged first. So do not merge it now. 
It is great to have it pre-reviewed to speed things up.

**Comment by nouiz** [comment]
I rebased this PR since the dependent PR is merged.
I also did the review comment. So it should be ready to merge now.

**Comment by nouiz** [comment]
Any update?

File: tensorflow/core/framework/allocator.h [hunk 2]
```diff
@@ -294,7 +294,7 @@ class Allocator {
 
   // For allocator that are stream aware, allow to specify the compute
   // stream this allocator is used for.
-  virtual void SetStream(void* stream) {}
+  virtual void SetStreamAndPreallocateMemory(void* stream) {}
```
Comment (sanjoy) [inline tensorflow/core/framework/allocator.h:297]: Update comment.

**Review by sanjoy — CHANGES_REQUESTED** [review]
Only a minor comment.

File: tensorflow/core/framework/allocator.h (hunk 2 above)
Comment (nouiz) [reply tensorflow/core/framework/allocator.h:297]: Done.

**Review by sanjoy — APPROVED** [review]

**Comment by nouiz** [comment]
The CI seems to have failed:
```
feedback/copybara — Google internal checks FAILED for runs with create time 2021-09-09T07:29:11.802148007Z.
```

Is this related to this PR?

**Pull request merged**